### ISSUE TITLE
TestBed should be zoneless compatible

### ISSUE THREAD

**Issue by yjaaidi** [angular/angular#48198]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

**TL;DR**: `TestBed` should provide a way to run zoneless and avoid importing `zone.js` when we don't need it.
And I'd love to help with a PR?

`TestBed` assumes that `zone.js` is available while there are contexts when this is not necessary and here are some examples:
- zoneless apps & libs
- tests that do not depend on change detection: service tests, pipe tests, "isolated" component testing etc...
- tests where CD is triggered manually using `fixture.detectChanges()`

These tests must load `zone.js` when they don't need it.
While this might sound like a really minor issue, loading `zone.js` can cause trouble in some situations.

For instance, integrating `TestBed` and `zone.js` with vitest causes the following error `Method Promise.prototype.then called on incompatible receiver` which is probably due to `zone.js` patching promises and breaking Node.js's safe promises when importing ESMs dynamically or something like that. This is similar to the issue described here https://github.com/angular/angular/issues/47872.

### Proposed solution

A first quick win is replacing this:
https://github.com/angular/angular/blob/6948f0f0f39c65f3dd0b2098774083f9c5a93352/packages/core/testing/src/test_bed_compiler.ts#L749-L751

with:
```ts
{provide: NgZone, useFactory: () => new NgZone({enableLongStackTrace: true}); }
```

this would then allow us to go zoneless with this configuration:

```ts
TestBed.configureTestingModule({
  providers: [
    {
      provide: NgZone,
      useClass: ɵNoopNgZone,
    },
  ],
});
```

The main problem with this approach is the boilerplate and having to use the `ɵNoopNgZone` implementation detail.

A second step would be adding the bootstrap's `ngZone` option to `initTestEnvironment`:
```ts
getTestBed().initTestEnvironment(
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting(),
  {ngZone: 'noop'}
);
``` 

### Alternatives considered

Disable promises monkey patching using `globalThis.__Zone_disable_ZoneAwarePromise = true`
but this causes the following issue https://github.com/angular/angular/issues/37349
Cf. https://github.com/angular/angular/issues/48359

**Comment by AndrewKushnir** [comment]
@yjaaidi thanks for submitting this ticket.

The first step of making the `new NgZone` creation lazy makes sense, please let us know if you are interested in submitting a PR. As you mentioned, that should provide an ability to replace ZoneJS implementation and experiment with vitest integration.

The second step of exposing a config option would require some discussions within the team. I'll share an update once I have it.

Thank you.

**Comment by yjaaidi** [comment]
Thanks @AndrewKushnir for your quick feedback.
This is great! And I'll be happy to help a PR pretty soon 😉

While investigating a bit further, I discovered another spot.
When there is a global `afterEach` (jasmine, jest, cypress, vitest with globals set to true), `TestBed`'s cleanup hook calls `resetFakeAsyncZone` which throws the following error.
https://github.com/angular/angular/blob/0ff5d972cdd20428d647a94c6c769c0759cfd49c/packages/core/testing/src/fake_async.ts#L25

IMO, this shouldn't throw an error. The error doesn't make much sense if we're not using `fakeAsync` and `fakeAsync` already throws the error.

Then the function should actually be named something like `tryResetFakeAsyncZone` but as the function is part of the public API we will have to keep it as is or alias it `resetFakeAsyncZone = tryResetFakeAsyncZone`.

**Comment by AndrewKushnir** [comment]
@yjaaidi the mentioned call is likely coming from here:
https://github.com/angular/angular/blob/4d398a0babbbf354a6ebdc64735ba70b953b3e3a/packages/core/testing/src/test_hooks.ts#L34-L42

We need to experiment with this more, but we can probably update it eventually to detect if we are in "zone-less" mode:
```
function getCleanupHook(expectedTeardownValue: boolean) {
  return () => {
    const testBed = TestBedImpl.INSTANCE;
    if (testBed.shouldTearDownTestingModule() === expectedTeardownValue) {
      const ngZone = testBed.get(NgZone); // access injector before it gets destroyed in the call below
      testBed.resetTestingModule();
      if (!(ngZone instanceof ɵNoopNgZone)) {
        resetFakeAsyncZone();
      }
    }
  };
}
```

**Comment by yjaaidi** [comment]
Thanks @AndrewKushnir ! In fact, this could be a quick win.
Though, I still think that throwing an error about fakeAsync while nobody is using it is misleading. What do you think?

If someone is using fakeAsync, the error is already thrown anyway.

**Comment by AndrewKushnir** [comment]
> If someone is using fakeAsync, the error is already thrown anyway.

@yjaaidi could you please let me know which error you refer to?

**Comment by yjaaidi** [comment]
@AndrewKushnir it's this one: https://github.com/angular/angular/blob/0ff5d972cdd20428d647a94c6c769c0759cfd49c/packages/core/testing/src/fake_async.ts#L51-L56

**Comment by LayZeeDK** [comment]
@yjaaidi 

In component tests, the `NgZone` is disabled by providing `true` for the value of `ComponentFixtureNoNgZone` as seen in https://github.com/angular/angular/blob/4e098fa8a7e7d176f4022831bc34eec6e2dfda3c/packages/core/test/application_ref_spec.ts#L610-L613

Example
```typescript
import { ComponentFixtureNoNgZone, TestBed } from '@angular/core/testing'
import { MyComponent } from './my.component';

it('supports zoneless', () => {
  TestBed.configureTestingModule({
    providers: [
      { provide: ComponentFixtureNoNgZone, useValue: true },
    ],
  });
  
  const fixture = TestBed.createComponent(MyComponent);
});
```

The tests for `ComponentFixtureNoNgZone` are at https://github.com/angular/angular/blob/8c04fa7852446238f72a5545575d0b9e66cf9b78/packages/core/test/component_fixture_spec.ts#L298-L329

**Comment by yjaaidi** [comment]
Hey @LayZeeDK !

Thanks for the help! Actually, `ComponentFixtureNoNgZone` skips injecting `NgZone` but `NgZone` was still instantiated and provided here (even if it's never injected nor used).  https://github.com/angular/angular/blob/6948f0f0f39c65f3dd0b2098774083f9c5a93352/packages/core/testing/src/test_bed_compiler.ts#L749-L751

But then it's already too late as the constructor of `NgZone` asserts the presence of Zone.js: https://github.com/angular/angular/blob/5e404d31143b6994b3a13ef7bb813e961dd2033a/packages/core/src/zone/ng_zone.ts#L128-L130

This made it mandatory to import `zone.js` even if it's not used... and that's what doesn't play well with Vitest & ESM in Node (due to promise monkey patching).

But guess what! Thanks to your comment, I checked again and this was indirectly fixed by the new `provideNgZoneChangeDetection()` function and mainly this commit https://github.com/angular/angular/commit/d7d6514add2912a18c50f190aaa8afafa313bc9e by @atscott 

Thanks so much to both of you!!!

**Comment by yjaaidi** [comment]
This was indirectly solved by https://github.com/angular/angular/pull/49557 by @atscott

**Comment by yjaaidi** [comment]
Also thank you @AndrewKushnir ! and sorry for the delay... I was off for some time and still catching up.
